bundle/edu.gemini.pot/src/main/scala/edu/gemini/spModel/gemini/ghost/GhostAsterism.scala

```diff
@@ -178,13 +182,13 @@ object GhostAsterism {
     }
   }
 
-  case class SingleTarget(target: GhostTarget, override val base: Option[Coordinates]) extends StandardResolution
-  case class DualTarget(target1: GhostTarget, target2: GhostTarget, override val base: Option[Coordinates]) extends StandardResolution
-  case class TargetPlusSky(target: GhostTarget, sky: Coordinates, override val base: Option[Coordinates]) extends StandardResolution
-  case class SkyPlusTarget(sky: Coordinates, target: GhostTarget, override val base: Option[Coordinates]) extends StandardResolution
+  case class SingleTarget(target: GhostTarget, override val base: Option[SPCoordinates]) extends StandardResolution
+  case class DualTarget(target1: GhostTarget, target2: GhostTarget, override val base: Option[SPCoordinates]) extends StandardResolution
+  case class TargetPlusSky(target: GhostTarget, sky: SPCoordinates, override val base: Option[SPCoordinates]) extends StandardResolution
+  case class SkyPlusTarget(sky: SPCoordinates, target: GhostTarget, override val base: Option[SPCoordinates]) extends StandardResolution
```

Comment:
The good news is also the bad news: `Asterism` was always mutable:

```
  // N.B. most members must be defs because `t` is mutable.
  final case class Single(t: SPTarget) extends Asterism {
    override def allSpTargets = NonEmptyList(t) // def because Nel isn't serializable
    override def allTargets = NonEmptyList(t.getTarget)
    override def basePosition(time: Option[Instant]) = t.getCoordinates(time.map(_.toEpochMilli))
    override def copyWithClonedTargets() = Single(t.clone)
    override def basePositionProperMotion = Target.pm.get(t.getTarget)
    override def asterismType: AsterismType = AsterismType.Single
  }
```
also `final case class GhostTarget(spTarget: SPTarget ...`

@sraaphorst will have all this fresh in his mind, but I believe it's impossible to get the OT and TPE working without rewriting them otherwise.

Reply:
Oh ok. I didn't look closely enough. All is "well"  👍

Reply:
As Shane said, `GhostAsterism` was always mutable since it has one or more `GhostTarget`s, each of which contain an `SPTarget`.

With regards to the `SPCoordinates`, the problem here was that we had two choices:

1. Use immutable `Coordinates` to represent sky positions, but then every time the RA / Dec of one of these changed in any way, we'd have to create a new `Asterism` and set it in the `TargetEnvironment`. This would have been a nightmare in terms of the target table and the TPE, since they rely on using `WatchablePos` with mutability to coordinate everything.

2. The easier alternative that fit in with how we're handling `SPTarget`s was to make a mutable `SPCoordinates` wrapper around `Coordinates` (kind of like how `SPTarget` wraps the three types of `Target`s). Then `SPCoordinates` could fit exactly into the TPE and target table using the existing `WatchablePos` / `TelescopePosWatcher` infrastructure.

By extracting the common `SPSkyObject` from `SPTarget` and `SPCoordinates`, I was able to reuse a lot of existing code that formerly would only work with `SPTarget` by parameterizing it `[T <: SPSkyObject]`.

Reply:
Ok, I understand. I stopped at this point when reviewing but I'll pick up there after lunch.